remove let and rewrite smacrolet using native let

allison-casey opened the pull request:
we can't use the `ScopeLet`'s `with` context manager to return a value in a regular `macro` in this case. Calling `compile` and returning the `Result` works fine, but may be bad practice? Honestly not sure, but it works and passes my tests so let me know if you have any thoughts.

Kodiologist commented:
Hmm, I think this approach is probably fine. Yes, it uses internal features of Hy, but so does some other stuff in Hyrule. It was never really a goal to make Hyrule only use Hy's public-facing bits.

`macrotools` doesn't make sense organizationally as a place to put this macro, because `smacrolet` is not intended primarily for writing other macros. Perhaps it should go in `misc`.

Don't forget the docstring.

I think `(. &compiler scope (create ScopeLet))` is better written `(.scope.create &compiler ScopeLet)`, and likewise ``(. &compiler (compile `(do ~@body)))`` as ``(.compile &compiler `(do ~@body))``, and `(scope.add …)` as `(.add scope …)`.

I suggest using `when` rather than `if` whenever you don't have an `else` form.

That's a clever use of `iter` with `zip`; it took me a bit to realize what you were doing. You can use `by2s` for this, though.

allison-casey commented:
* Rewrote the dot forms
* changed `if` to `when`
* Couldn't use `by2s` because of circular import to `let` module. Moved `by2s` to  `collections` (which makes more sense anyways) to solve this and moved `smacrolet` to `utils`

Kodiologist commented:
I rebased to remove Hyrule's `let` and remaining support code, split your work into two commits, and remove a stray edit in `test_macrotools.hy`.
